File: kafka-test-rust/src/producer.rs
use apache_avro::Schema;
use apache_avro::types::Record;
use rdkafka::ClientConfig;
use rdkafka::producer::{FutureProducer, FutureRecord};
use rdkafka::util::Timeout;
use std::env;
use std::fs::File;
use std::io::Read;
use std::path::Path;
use std::time::{Duration, SystemTime, UNIX_EPOCH};
use tokio::time;

fn load_schema() -> Schema {
    let paths = [
        "../schemas/message.avsc",
        "/app/schemas/message.avsc",
        "schemas/message.avsc", // Fallback
    ];

    for p in paths {
        if Path::new(p).exists() {
            let mut file = File::open(p).expect("Failed to open schema file");
            let mut content = String::new();
            file.read_to_string(&mut content)
                .expect("Failed to read schema file");
            return Schema::parse_str(&content).expect("Failed to parse schema");
        }
    }
    panic!("Schema file not found in paths: {:?}", paths);
}

#[tokio::main]
async fn main() {
    tracing_subscriber::fmt::init();
    tracing::info!("Hello from kafka-test-rust!");

    let schema = load_schema();
    tracing::info!("Loaded Avro schema");

    let bootstrap_servers =
        env::var("KAFKA_BOOTSTRAP_SERVERS").unwrap_or_else(|_| "127.0.0.1:9094".to_string());
    let topic = env::var("TOPIC_NAME").unwrap_or_else(|_| "my-topic-1".to_string());

    let producer: FutureProducer = ClientConfig::new()
        .set("bootstrap.servers", &bootstrap_servers)
        .set("message.timeout.ms", "5000")
        .set("broker.address.family", "v4")
        .create()
        .expect("Producer creation error");

    tracing::info!("🚀 Producer is now running...");

    let mut message_id = 0;
    loop {
        let event_time = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .expect("Time went backwards")
            .as_secs_f64();

        // Create Avro record manually since we don't have generated structs
        // Alternatively we could define a struct with Serde and use to_avro_datum
        let mut record = Record::new(&schema).unwrap();
        record.put("message_id", message_id);
        record.put("event_time", event_time);
        record.put("content", format!("Message {}", message_id));

        let payload =
            apache_avro::to_avro_datum(&schema, record).expect("Failed to serialize to Avro");

        // Send message
        match producer
            .send(
                FutureRecord::<(), [u8]>::to(&topic).payload(&payload),
                Timeout::Never,
            )
            .await
        {
            Ok(_) => tracing::info!("🚀 Sent: Message {} ({} bytes)", message_id, payload.len()),
            Err(e) => tracing::error!("❌ Error sending message: {:?}", e),
        }

        message_id += 1;

        // Wait with cancellation support
        tokio::select! {
            _ = time::sleep(Duration::from_secs(2)) => {}
            _ = tokio::signal::ctrl_c() => {
                tracing::info!("🛑 Shutting down producer...");
                break;
            }
        }
    }
}

File: kafka-test-ruby/producer.rb
require "rdkafka"
require "json"
require "time"
require "logger"
require "avro"
require "stringio"

def load_schema
  paths = [
    "../schemas/message.avsc",
    "/app/schemas/message.avsc",
    "schemas/message.avsc"
  ]

  path = paths.find { |p| File.exist?(p) }
  raise "Schema file not found in paths: #{paths}" unless path

  Avro::Schema.parse(File.read(path))
end

def main
  STDOUT.sync = true
  logger = Logger.new(STDOUT)
  logger.formatter = proc do |severity, datetime, progname, msg|
    "#{datetime.strftime('%Y-%m-%d %H:%M:%S')} #{severity} #{msg}\n"
  end

  logger.info "Hello from kafka-test-ruby!"

  schema = load_schema
  logger.info "Loaded Avro schema: #{schema.name}"

  bootstrap_servers = ENV.fetch("KAFKA_BOOTSTRAP_SERVERS", "127.0.0.1:9094")
  topic = ENV.fetch("TOPIC_NAME", "my-topic-1")

  config = {
    "bootstrap.servers": bootstrap_servers,
    "broker.address.family": "v4",
    "message.timeout.ms": 5000
  }

  producer = Rdkafka::Config.new(config).producer

  logger.info "🚀 Producer is now running..."

  message_id = 0
  writer = Avro::IO::DatumWriter.new(schema)

  loop do
    event_time = Time.now.to_f
    message = {
      "message_id" => message_id,
      "event_time" => event_time,
      "content" => "Message #{message_id}"
    }

    buffer = StringIO.new
    encoder = Avro::IO::BinaryEncoder.new(buffer)
    writer.write(message, encoder)
    payload = buffer.string

    begin
      producer.produce(
          topic: topic,
          payload: payload
      ).wait

      logger.info "🚀 Sent: #{message}"
    rescue Rdkafka::RdkafkaError => e
      logger.error "❌ Error sending message: #{e.message}"
    end

    message_id += 1
    sleep 2
  end
rescue Interrupt
  logger.info "🛑 Shutting down producer..."
ensure
  if defined?(producer) && producer
    producer.close
    logger.info "✅ Producer closed."
  end
end

main if __FILE__ == $0
